File: ruby/clock/clock.rb
class Clock
  attr_reader :hour, :minute

  def initialize(hour: 0, minute: 0)
    @hour = hour
    @minute = minute

    div, mod = @minute.divmod(60)
    @hour += div
    @minute = mod

    @hour %= 24
  end

  def ==(other)
    @hour == other.hour && @minute == other.minute
  end

  def +(other)
    Clock.new(hour: @hour + other.hour, minute: @minute + other.minute)
  end

  def -(other)
    Clock.new(hour: @hour - other.hour, minute: @minute - other.minute)
  end

  def to_s
    "%02i:%02i" % [@hour, @minute]
  end
end

File: cpp/clock/clock.cpp
#include "clock.h"

using namespace std;

date_independent::clock::clock(int h, int m) {
    hour = h;
    minute = m;

    while (minute >= 60) {
        minute -= 60;
        hour += 1;
    }

    while (minute < 0) {
        minute += 60;
        hour -= 1;
    }

    if (hour >= 24)
        hour %= 24;

    while (hour < 0)
        hour += 24;
}

date_independent::clock date_independent::clock::at(int h, int m) {
    return clock(h, m);
}

date_independent::clock::operator string() const {
    auto hh = to_string(hour);
    auto mm = to_string(minute);

    if (hour < 10)
        hh.insert(hh.begin(), 1, '0');

    if (minute < 10)
        mm.insert(mm.begin(), 1, '0');

    return hh + ":" + mm;
}

date_independent::clock date_independent::clock::plus(int n) {
    return clock(hour, minute + n);
}

bool date_independent::clock::operator==(const date_independent::clock& other) const {
    return hour == other.get_hours() && minute == other.get_minutes();
}

bool date_independent::clock::operator!=(const date_independent::clock& other) const {
    return !(*this == other);
}

int date_independent::clock::get_hours() const {
    return hour;
}

int date_independent::clock::get_minutes() const {
    return minute;
}
